### ISSUE TITLE
Feature Request: Composition Events (Keyboard API)

### ISSUE THREAD

**Issue by ellatrix** [puppeteer/puppeteer#4981]
We're using Puppeteer at WordPress, where we've created a custom editor. One thing we keep running into is bugs around input composition. Every now and then things break, and we're not able to create tests for it with Puppeteer. It would be great if Puppeteer provided an additional `keyboard` API to simulate composing characters with the keyboard.

https://w3c.github.io/uievents/#events-composition-types

In other words, this will not produce `è` in Puppeteer:

```js
await page.keyboard.down( 'Alt' );
await page.keyboard.type( '`' );
await page.keyboard.up( 'Alt' );
await page.keyboard.type( 'e' );
```

Thanks for the consideration!

**Comment by SProst** [comment]
@ellatrix CDP has experimental support for IME https://chromedevtools.github.io/devtools-protocol/tot/Input#method-insertText. Have you tried using this function directly to simulate composing characters?

**Comment by ellatrix** [comment]
@SProst I have not! Thanks for the tip!

**Comment by ellatrix** [comment]
Unfortunately, that doesn't work for testing composition. There's only one method to immediately insert all text. I'm looking for something to do this step-by-step. So insert the first character of a composition, but don't end the composition, then the second, then do something else, then the third... etc.

Here's a use case: I'd like to ensure we don't interfere in any way with the user composing characters. If we would reset the DOM during composition, this special state is destroyed. The state I'm talking about is visible with an underline.

<img width="49" alt="Screenshot 2019-08-08 at 11 51 01" src="https://user-images.githubusercontent.com/4710635/65828921-dd7f4780-e2a8-11e9-8e56-ba2c568f43a7.png">

In the screenshot above you can see that I'm composing some text after typing a back tick (`` ` ``), because it's underlined by the browser. You can even do this with an American keyboard by pressing `` Alt+` `` to compose a vowel with an accent, for example `è`. Press `` Alt+` `` followed by `e` to create `è`.

There's no way I can test this step-by-step in puppeteer. I've tried to dispatch all the events normally fired by Chrome during composition, but I cannot recreate the special state that Chrome is is that would compose `è` at the end of the sequence.

Thanks for your help!

**Comment by ellatrix** [comment]
So, in other words, this will not produce `è` in Puppeteer:

```js
await page.keyboard.down( 'Alt' );
await page.keyboard.type( '`' );
await page.keyboard.up( 'Alt' );
await page.keyboard.type( 'e' );
```

**Comment by SProst** [comment]
@ellatrix What you have described might need to be either a feature enhancement to the CDP or (likely) a feature change to Chromium that then translates into an enhancement to the CDP. @aslushnikov Thoughts?

**Comment by mrkev** [comment]
Hello! I help maintain https://github.com/facebook/draft-js and this would be huge in helping us build it with confidence. Currently we have no way to automatically test if a change to fix, say, selection in English ends up breaking input composition in Chinese.

Commenting to show our interest and listen in on any updates.

**Comment by vocafeuvre** [comment]
good news, just found a page about a CDP method called 'Input.imeSetComposition' in Chromium Code Review. once this feature is shipped, it might resolve the issue about lack of support upstream. the question now is would Puppeteer be implementing a wrapper around this method?

sources:
https://github.com/microsoft/playwright/issues/5777
https://chromium-review.googlesource.com/c/chromium/src/+/2764592

**Comment by mrkev** [comment]
I no longer work on draftjs, but I'm happy this is moving forward 👍

**Comment by ellatrix** [comment]
This is still an issue. Just preventing @stale from closing it.
